[MegaCache] Encode key in base64

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151472

I have noticed that there are some errors like
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x95 in position 169302: invalid start byte
```

I havent been able to repro this locally yet, this change should fix the encoding issues

oulgen committed:
[MegaCache] Encode key in base64

I have noticed that there are some errors like
"""
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x95 in position 169302: invalid start byte
"""

I havent been able to repro this locally yet, this change should fix the encoding issues

[ghstack-poisoned]

oulgen committed:
None on "[MegaCache] Encode key in base64"


I have noticed that there are some errors like
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x95 in position 169302: invalid start byte
```

I havent been able to repro this locally yet, this change should fix the encoding issues

[ghstack-poisoned]

masnesral commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -49,7 +51,7 @@ def read_uint64(self) -> int:
 
     def read_str(self) -> str:
-        return self.read_bytes().decode("utf-8")
+        return base64.b64decode(self.read_bytes()).decode("utf-8")
```
Is the paren in the right place? decode first and then b64decode?

masnesral replied:
`base64.b64decode(self.read_bytes().decode("utf-8"))`?

masnesral replied:
What am I saying? I think it's right.

aorenste commented:
Why do we think this will help? If `write_bytes` and `read_bytes` aren't reading and writing bytes we've got a problem. And if it's chopping the strings then base64 isn't going to be better because it also depends on multi-byte encodings...

oulgen committed:
None on "[MegaCache] Encode key in base64"


I have noticed that there are some errors like
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x95 in position 169302: invalid start byte
```

I havent been able to repro this locally yet, this change should fix the encoding issues

[ghstack-poisoned]

oulgen committed:
None on "[MegaCache] Encode key in base64"


I have noticed that there are some errors like
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x95 in position 169302: invalid start byte
```

I havent been able to repro this locally yet, this change should fix the encoding issues

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge -f "it says i can merge successfully"
